File: src/core/sync-state.ts
/**
 * Project-local sync state for near-seamless switching.
 *
 * xfer intentionally keeps sync metadata out of Claude/Codex internals. The map
 * only records which source session has already been migrated to which target
 * session for a given cwd, so `switch` can reuse existing destinations instead
 * of duplicating sessions every time.
 */

import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import type { Agent, SessionIndexEntry } from "./model.js";
import { configuredStatePath } from "./settings.js";
import { normalizeProjectCwd } from "./paths.js";

export interface SessionMapEntry {
  cwd: string;
  sourceAgent: Agent;
  sourceSessionId: string;
  targetAgent: Agent;
  targetSessionId: string;
  targetFilePath: string;
  createdAt: string;
  sourceUpdatedAt?: string;
}

export interface SyncState {
  version: 1;
  mappings: SessionMapEntry[];
}

export function statePath(cwd = process.cwd()): string {
  return configuredStatePath(cwd) || path.join(cwd, ".xfer", "state.json");
}

export function loadState(cwd = process.cwd()): SyncState {
  const file = statePath(cwd);
  try {
    return JSON.parse(fs.readFileSync(file, "utf8")) as SyncState;
  } catch {
    return { version: 1, mappings: [] };
  }
}

export function saveState(state: SyncState, cwd = process.cwd()): void {
  const file = statePath(cwd);
  fs.mkdirSync(path.dirname(file), { recursive: true });
  const tmp = path.join(path.dirname(file), `.state-${process.pid}-${Date.now()}.tmp`);
  fs.writeFileSync(tmp, JSON.stringify(state, null, 2) + "\n", "utf8");
  fs.renameSync(tmp, file);
}

export function findMapping(
  state: SyncState,
  sourceAgent: Agent,
  sourceSessionId: string,
  targetAgent: Agent
): SessionMapEntry | undefined {
  return state.mappings.find(
    (m) =>
      m.sourceAgent === sourceAgent &&
      m.sourceSessionId === sourceSessionId &&
      m.targetAgent === targetAgent
  );
}

export function upsertMapping(
  state: SyncState,
  entry: SessionMapEntry
): SyncState {
  const next = state.mappings.filter(
    (m) =>
      !(
        m.sourceAgent === entry.sourceAgent &&
        m.sourceSessionId === entry.sourceSessionId &&
        m.targetAgent === entry.targetAgent
      )
  );
  next.push(entry);
  next.sort((a, b) => b.createdAt.localeCompare(a.createdAt));
  return { version: 1, mappings: next };
}

export function latestForCwd(
  sessions: SessionIndexEntry[],
  cwd: string,
  agent?: Agent
): SessionIndexEntry | undefined {
  const target = normalizeCwd(cwd);
  return sessions
    .filter((s) => (!agent || s.agent === agent) && s.cwd && normalizeCwd(s.cwd) === target)
    .sort((a, b) => (b.updatedAt || "").localeCompare(a.updatedAt || ""))[0];
}

export function otherAgent(agent: Agent): Agent {
  return agent === "claude" ? "codex" : "claude";
}

export function normalizeCwd(p: string): string {
  return normalizeProjectCwd(p).replace(/[\\/]+/g, "/").replace(/\/$/, "").toLowerCase();
}

export function defaultProjectCwd(cwd = process.cwd()): string {
  return normalizeProjectCwd(path.resolve(cwd));
}

File: src-tauri/src/xfer/sync_state.rs
use crate::xfer::db;
use crate::xfer::model::{Agent, SessionIndexEntry};
use crate::xfer::paths::{default_project_cwd, normalize_cwd as normalize_project_key, normalize_project_cwd};
use crate::xfer::settings::configured_state_path;
use serde::{Deserialize, Serialize};
use std::path::PathBuf;

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionMapEntry {
  pub cwd: String,
  pub source_agent: Agent,
  pub source_session_id: String,
  pub target_agent: Agent,
  pub target_session_id: String,
  pub target_file_path: String,
  pub created_at: String,
  pub source_updated_at: Option<String>,
}

#[derive(Clone, Deserialize, Serialize)]
pub struct SyncState {
  pub version: u8,
  pub mappings: Vec<SessionMapEntry>,
}

pub fn state_path(cwd: &str) -> PathBuf {
  configured_state_path(cwd).unwrap_or_else(|| PathBuf::from(cwd).join(".xfer").join("state.json"))
}

pub fn load_state(cwd: &str) -> SyncState {
  let file = state_path(cwd);
  let mut state = std::fs::read_to_string(file)
    .ok()
    .and_then(|raw| serde_json::from_str(&raw).ok())
    .unwrap_or(SyncState { version: 1, mappings: Vec::new() });
  if let Ok(mappings) = db::list_mappings(cwd) {
    state = merge_mappings(state, mappings);
  }
  state
}

pub fn save_state(state: &SyncState, cwd: &str) -> Result<(), String> {
  for mapping in &state.mappings {
    let _ = db::upsert_mapping(mapping);
  }
  let file = state_path(cwd);
  let dir = file.parent().ok_or_else(|| format!("invalid state path: {}", file.display()))?;
  std::fs::create_dir_all(dir).map_err(|e| e.to_string())?;
  let tmp = dir.join(format!(".state-{}-{}.tmp", std::process::id(), timestamp_millis()));
  std::fs::write(&tmp, serde_json::to_string_pretty(state).map_err(|e| e.to_string())? + "\n")
    .map_err(|e| e.to_string())?;
  std::fs::rename(tmp, file).map_err(|e| e.to_string())
}

fn merge_mappings(mut state: SyncState, mappings: Vec<SessionMapEntry>) -> SyncState {
  for entry in mappings {
    state = upsert_mapping(&state, entry);
  }
  state
}

pub fn find_mapping<'a>(
  state: &'a SyncState,
  source_agent: Agent,
  source_session_id: &str,
  target_agent: Agent,
) -> Option<&'a SessionMapEntry> {
  state.mappings.iter().find(|m| {
    m.source_agent == source_agent
      && m.source_session_id == source_session_id
      && m.target_agent == target_agent
  })
}

pub fn upsert_mapping(state: &SyncState, entry: SessionMapEntry) -> SyncState {
  let entry_cwd = normalize_project_key(&entry.cwd);
  let mut next: Vec<SessionMapEntry> = state
    .mappings
    .iter()
    .cloned()
    .filter(|m| {
      !(normalize_project_key(&m.cwd) == entry_cwd
        && m.source_agent == entry.source_agent
        && m.source_session_id == entry.source_session_id
        && m.target_agent == entry.target_agent)
    })
    .collect();
  next.push(entry);
  next.sort_by(|a, b| b.created_at.cmp(&a.created_at));
  SyncState { version: 1, mappings: next }
}

pub fn latest_for_cwd(
  sessions: &[SessionIndexEntry],
  cwd: &str,
  agent: Agent,
) -> Option<SessionIndexEntry> {
  let target = normalize_cwd(cwd);
  sessions
    .iter()
    .filter(|s| s.agent == agent)
    .filter(|s| s.cwd.as_deref().map(normalize_cwd) == Some(target.clone()))
    .cloned()
    .max_by(|a, b| a.updated_at.cmp(&b.updated_at))
}

pub fn other_agent(agent: Agent) -> Agent {
  agent.other()
}

pub fn resume_command(agent: Agent, session_id: &str) -> String {
  match agent {
    Agent::Claude => format!("claude --resume {session_id}"),
    Agent::Codex => format!("codex resume {session_id}"),
  }
}

pub fn default_cwd(cwd: Option<&str>) -> String {
  default_project_cwd(cwd)
}

fn normalize_cwd(p: &str) -> String {
  normalize_project_cwd(p).replace(['\\', '/'], "/").trim_end_matches('/').to_lowercase()
}

fn timestamp_millis() -> u128 {
  use std::time::{SystemTime, UNIX_EPOCH};
  SystemTime::now()
    .duration_since(UNIX_EPOCH)
    .map(|d| d.as_millis())
    .unwrap_or(0)
}
